source/common/upstream/upstream_impl.cc

```diff
@@ -445,13 +462,50 @@ void PrioritySetImpl::updateHosts(uint32_t priority, UpdateHostsParams&& update_
                                   absl::optional<uint32_t> overprovisioning_factor) {
   // Ensure that we have a HostSet for the given priority.
   getOrCreateHostSet(priority, overprovisioning_factor);
-  // TODO(snowp): Add a batched update mode that allows updating multiple HostSet and invoke the
-  // membership update cb for the resulting host diff.
   static_cast<HostSetImpl*>(host_sets_[priority].get())
       ->updateHosts(std::move(update_hosts_params), std::move(locality_weights), hosts_added,
                     hosts_removed, overprovisioning_factor);
 
-  runUpdateCallbacks(hosts_added, hosts_removed);
+  if (!batch_update_) {
+    runUpdateCallbacks(hosts_added, hosts_removed);
+  }
+}
+
+void PrioritySetImpl::batchHostUpdate(std::function<void(UpdateHostsCb)> callback) {
+  ASSERT(!batch_update_);
+  batch_update_ = true;
+
+  std::unordered_set<uint32_t> priorities;
+  std::unordered_set<HostSharedPtr> all_hosts_added;
+  std::unordered_set<HostSharedPtr> all_hosts_removed;
+
+  // We wrap the update call with a lambda that tracks all the hosts that have been added/removed.
+  callback([&](auto p, auto&& params, auto locality_weight, const auto& hosts_added,
+               const auto& hosts_removed, auto overprovisioning_factor) {
+    // We assume that each call updates a different priority.
+    ASSERT(priorities.find(p) == priorities.end());
+    priorities.insert(p);
+
+    for (const auto& host : hosts_added) {
+      all_hosts_added.insert(host);
+    }
+
+    for (const auto& host : hosts_removed) {
+      all_hosts_removed.insert(host);
+    }
+
+    updateHosts(p, std::move(params), locality_weight, hosts_added, hosts_removed,
+                overprovisioning_factor);
+  });
+
+  // TODO(snowp): Use RAII to make this exception safe?
```

Comment:
Can we get exceptions during updates? I.e. have we done all the validation at this point? If we can, we probably need this.

Reply:
I don't think we throw anything right now, but it kinda feels like we should do this now in case that ever changes in the future